[cc65] Declarations of variables of incomplete types?

So... anyone enlighten me please?
```c
struct S s; /* Question: is this allowed? GCC: Yes; Clang: Yes; MSVC: No;  CC65: No */
enum E e;   /* Question: is this allowed? GCC: No;  Clang: Yes; MSVC: Yes; CC65: Yes (No with my #1093 solution) */

struct S {
    int a;
};

enum E {
    Z
};
```

````
1118.c:2:6: warning: ISO C forbids forward references to 'enum' types [-Wpedantic]
    2 | enum E e;   /* Question: is this allowed? GCC: No;  Clang: Yes; MSVC: Yes; CC65: Yes (No with my #1093 solution) */
      |      ^
````
vbcc: Yes
vbcc: Yes

--- a/src/cc65/compile.c
+++ b/src/cc65/compile.c
@@ -415,9 +415,36 @@ void Compile (const char* FileName)
 
     } else {
 
+        /* Used for emitting externals */
+        SymEntry* Entry;
+
         /* Ok, start the ball rolling... */
         Parse ();
 
+        /* Reset the BSS segment name to its default; so that the below strcmp()
+        ** will work as expected, at the beginning of the list of variables
+        */
+        SetSegName (SEG_BSS, SEGNAME_BSS);
+
+        /* Walk over all global symbols and generate code for uninitialized
+        ** global variables.
+        */
+        for (Entry = GetGlobalSymTab ()->SymHead; Entry; Entry = Entry->NextSym) {
+            if ((Entry->Flags & (SC_STORAGE | SC_DEF | SC_STATIC)) == (SC_STORAGE | SC_STATIC)) {
+                /* Assembly definition of uninitialized global variable */
+
+                /* Set the segment name only when it changes */
+                if (strcmp (GetSegName (SEG_BSS), Entry->V.BssName) != 0) {
+                    SetSegName (SEG_BSS, Entry->V.BssName);
+                    g_segname (SEG_BSS);
+                }
+                g_usebss ();
+                g_defgloblabel (Entry->Name);
+                g_res (SizeOf (Entry->Type));
+                /* Mark as defined; so that it will be exported, not imported */
+                Entry->Flags |= SC_DEF;
+            }
+        }
     }
 
     if (Debug) {
@@ -431,18 +458,12 @@ void Compile (const char* FileName)
 
 
 void FinishCompile (void)
-/* Emit literals, externals, debug info, do cleanup and optimizations */
+/* Emit literals, debug info, do cleanup and optimizations */
 {
     SymEntry* Entry;
 
-    /* Reset the BSS segment name to its default; so that the below strcmp()
-    ** will work as expected, at the beginning of the list of variables
-    */
-    SetSegName (SEG_BSS, SEGNAME_BSS);
-
-    /* Walk over all global symbols:
-    ** - for functions, do clean-up and optimizations
-    ** - generate code for uninitialized global variables
+    /* Walk over all global symbols and do clean-up and optimizations for
+    ** functions.
     */
     for (Entry = GetGlobalSymTab ()->SymHead; Entry; Entry = Entry->NextSym) {
         if (SymIsOutputFunc (Entry)) {
@@ -450,19 +471,6 @@ void FinishCompile (void)
             MoveLiteralPool (Entry->V.F.LitPool);
             CS_MergeLabels (Entry->V.F.Seg->Code);
             RunOpt (Entry->V.F.Seg->Code);
-        } else if ((Entry->Flags & (SC_STORAGE | SC_DEF | SC_STATIC)) == (SC_STORAGE | SC_STATIC)) {
-            /* Assembly definition of uninitialized global variable */
-
-            /* Set the segment name only when it changes */
-            if (strcmp (GetSegName (SEG_BSS), Entry->V.BssName) != 0) {
-                SetSegName (SEG_BSS, Entry->V.BssName);
-                g_segname (SEG_BSS);
-            }
-            g_usebss ();
-            g_defgloblabel (Entry->Name);
-            g_res (SizeOf (Entry->Type));
-            /* Mark as defined; so that it will be exported, not imported */
-            Entry->Flags |= SC_DEF;
         }
     }
 

--- a/src/cc65/compile.c
+++ b/src/cc65/compile.c
@@ -449,7 +449,7 @@ void Compile (const char* FileName)
                         /* Assume array size of 1 */
                         SetElementCount (Entry->Type, 1);
                         Size = SizeOf (Entry->Type);
-                        Warning ("Tentative array '%s[]' assumed to have one element", Entry->Name);
+                        Warning ("Incomplete array '%s[]' assumed to have one element", Entry->Name);
                     }
 
                     Sym = GetSymType (GetElementType (Entry->Type));
@@ -474,7 +474,7 @@ void Compile (const char* FileName)
                     Entry->Flags |= SC_DEF;
                 } else {
                     /* Tentative declared variable is still of incomplete type */
-                    Error ("Tentative definition of '%s' of type '%s' is incomplete",
+                    Error ("Definition of '%s' has type '%s' that is never completed",
                            Entry->Name,
                            GetFullTypeName (Entry->Type));
                 }

--- a/src/cc65/compile.c
+++ b/src/cc65/compile.c
@@ -148,6 +148,11 @@ static void Parse (void)
                 break;
             }
 
+            if ((Decl.StorageClass & SC_ALIAS) == SC_ALIAS) {
+                /* Failed parsing */
+                goto SkipOneDecl;
+            }
+
             /* Check if we must reserve storage for the variable. We do this,
             **
             **   - if it is not a typedef or function,
@@ -276,6 +281,7 @@ static void Parse (void)
 
             }
 
+SkipOneDecl:
             /* Check for end of declaration list */
             if (CurTok.Tok == TOK_COMMA) {
                 NextToken ();

--- a/src/cc65/compile.c
+++ b/src/cc65/compile.c
@@ -248,15 +248,10 @@ static void Parse (void)
                         /* We cannot declare variables of type void */
                         Error ("Illegal type for variable '%s'", Decl.Ident);
                         Entry->Flags &= ~(SC_STORAGE | SC_DEF);
-                    } else if (Size == 0) {
+                    } else if (Size == 0 && SymIsDef (Entry)) {
                         /* Size is unknown. Is it an array? */
                         if (!IsTypeArray (Decl.Type)) {
                             Error ("Variable '%s' has unknown size", Decl.Ident);
-                        }
-                        /* Do this only if the same array has not been defined */
-                        if (!SymIsDef (Entry)) {
-                            Entry->Flags &= ~(SC_STORAGE | SC_DEF);
-                            Entry->Flags |= SC_DECL;
                         }
                     } else {
                         /* A global (including static) uninitialized variable is
@@ -432,17 +427,42 @@ void Compile (const char* FileName)
         for (Entry = GetGlobalSymTab ()->SymHead; Entry; Entry = Entry->NextSym) {
             if ((Entry->Flags & (SC_STORAGE | SC_DEF | SC_STATIC)) == (SC_STORAGE | SC_STATIC)) {
                 /* Assembly definition of uninitialized global variable */
+                SymEntry* Sym = GetSymType (Entry->Type);
+                unsigned Size = SizeOf (Entry->Type);
+                if (Size == 0 && IsTypeArray (Entry->Type)) {
+                    if (GetElementCount (Entry->Type) == UNSPECIFIED) {
+                        /* Assume array size of 1 */
+                        SetElementCount (Entry->Type, 1);
+                        Size = SizeOf (Entry->Type);
+                        Warning ("Tentative array '%s[]' assumed to have one element", Entry->Name);
+                    }
 
-                /* Set the segment name only when it changes */
-                if (strcmp (GetSegName (SEG_BSS), Entry->V.BssName) != 0) {
-                    SetSegName (SEG_BSS, Entry->V.BssName);
-                    g_segname (SEG_BSS);
+                    Sym = GetSymType (GetElementType (Entry->Type));
+                    if (Size == 0 && Sym != 0 && SymIsDef (Sym)) {
+                        /* Array of 0-size elements */
+                        Warning ("Array '%s[]' has 0-sized elements", Entry->Name);
+                    }
                 }
-                g_usebss ();
-                g_defgloblabel (Entry->Name);
-                g_res (SizeOf (Entry->Type));
-                /* Mark as defined; so that it will be exported, not imported */
-                Entry->Flags |= SC_DEF;
+
+                /* For non-ESU types, Size != 0 */
+                if (Size != 0 || (Sym != 0 && SymIsDef (Sym))) {
+                    /* Set the segment name only when it changes */
+                    if (strcmp (GetSegName (SEG_BSS), Entry->V.BssName) != 0) {
+                        SetSegName (SEG_BSS, Entry->V.BssName);
+                        g_segname (SEG_BSS);
+                    }
+                    g_usebss ();
+                    g_defgloblabel (Entry->Name);
+                    g_res (Size);
+
+                    /* Mark as defined; so that it will be exported, not imported */
+                    Entry->Flags |= SC_DEF;
+                } else {
+                    /* Tentative declared variable is still of incomplete type */
+                    Error ("Tentative definition of '%s' of type '%s' is incomplete",
+                           Entry->Name,
+                           GetFullTypeName (Entry->Type));
+                }
             }
         }
     }

--- a/src/cc65/compile.c
+++ b/src/cc65/compile.c
@@ -276,6 +276,15 @@ static void Parse (void)
                                    Entry->Name, Entry->V.BssName);
                         }
                         Entry->V.BssName = xstrdup (bssName);
+
+                        /* Check for enum forward declaration.
+                        ** Warn about it when extensions are not allowed.
+                        */
+                        if (Size == 0 && IsTypeEnum (Decl.Type)) {
+                            if (IS_Get (&Standard) != STD_CC65) {
+                                Warning ("ISO C forbids forward references to 'enum' types");
+                            }
+                        }
                     }
                 }
 

--- a/src/cc65/compile.h
+++ b/src/cc65/compile.h
@@ -48,7 +48,7 @@ void Compile (const char* FileName);
 /* Top level compile routine. Will setup things and call the parser. */
 
 void FinishCompile (void);
-/* Emit literals, externals, do cleanup and optimizations */
+/* Emit literals, debug info, do cleanup and optimizations */
 
 
 

--- a/src/cc65/declare.c
+++ b/src/cc65/declare.c
@@ -920,7 +920,7 @@ NextMember: if (CurTok.Tok != TOK_COMMA) {
     LeaveStructLevel ();
 
     /* Make a real entry from the forward decl and return it */
-    return AddStructSym (Name, SC_UNION, UnionSize, FieldTab);
+    return AddStructSym (Name, SC_UNION | SC_DEF, UnionSize, FieldTab);
 }
 
 
@@ -1102,7 +1102,7 @@ NextMember: if (CurTok.Tok != TOK_COMMA) {
     LeaveStructLevel ();
 
     /* Make a real entry from the forward decl and return it */
-    return AddStructSym (Name, SC_STRUCT, StructSize, FieldTab);
+    return AddStructSym (Name, SC_STRUCT | SC_DEF, StructSize, FieldTab);
 }
 
 

--- a/src/cc65/declare.c
+++ b/src/cc65/declare.c
@@ -1814,6 +1814,9 @@ Type* ParseType (Type* T)
 void ParseDecl (const DeclSpec* Spec, Declaration* D, declmode_t Mode)
 /* Parse a variable, type or function declaration */
 {
+    /* Used to check if we have any errors during parsing this */
+    unsigned PrevErrorCount = ErrorCount;
+
     /* Initialize the Declaration struct */
     InitDeclaration (D);
 
@@ -1891,8 +1894,8 @@ void ParseDecl (const DeclSpec* Spec, Declaration* D, declmode_t Mode)
         }
     }
 
-    /* Check the size of the generated type */
     if (!IsTypeFunc (D->Type) && !IsTypeVoid (D->Type)) {
+        /* Check the size of the generated type */
         unsigned Size = SizeOf (D->Type);
         if (Size >= 0x10000) {
             if (D->Ident[0] != '\0') {
@@ -1901,8 +1904,12 @@ void ParseDecl (const DeclSpec* Spec, Declaration* D, declmode_t Mode)
                 Error ("Invalid size in declaration (0x%06X)", Size);
             }
         }
-    }
 
+        if (PrevErrorCount != ErrorCount) {
+            /* Don't give storage if the declaration is not parsed correctly */
+            D->StorageClass |= SC_DECL | SC_ALIAS;
+        }
+    }
 }
 
 

--- a/src/cc65/expr.c
+++ b/src/cc65/expr.c
@@ -244,6 +244,23 @@ static int TypeSpecAhead (void)
 
 
 
+static unsigned ExprCheckedSizeOf (const Type* T)
+/* Specially checked SizeOf() used in 'sizeof' expressions */
+{
+    unsigned Size = SizeOf (T);
+    SymEntry* Sym;
+
+    if (Size == 0) {
+        Sym = GetSymType (T);
+        if (Sym == 0 || !SymIsDef (Sym)) {
+            Error ("Cannot apply 'sizeof' to incomplete type '%s'", GetFullTypeName (T));
+        }
+    }
+    return Size;
+}
+
+
+
 void PushAddr (const ExprDesc* Expr)
 /* If the expression contains an address that was somehow evaluated,
 ** push this address on the stack. This is a helper function for all
@@ -1890,7 +1907,7 @@ void hie10 (ExprDesc* Expr)
             if (TypeSpecAhead ()) {
                 Type T[MAXTYPELEN];
                 NextToken ();
-                Size = CheckedSizeOf (ParseType (T));
+                Size = ExprCheckedSizeOf (ParseType (T));
                 ConsumeRParen ();
             } else {
                 /* Remember the output queue pointer */
@@ -1908,7 +1925,7 @@ void hie10 (ExprDesc* Expr)
                         ReleaseLiteral (Expr->LVal);
                     }
                     /* Calculate the size */
-                    Size = CheckedSizeOf (Expr->Type);
+                    Size = ExprCheckedSizeOf (Expr->Type);
                 }
                 /* Remove any generated code */
                 RemoveCode (&Mark);

--- a/src/cc65/locals.c
+++ b/src/cc65/locals.c
@@ -443,13 +443,14 @@ static void ParseOneDecl (const DeclSpec* Spec)
     }
 
     /* If the symbol is not marked as external, it will be defined now */
-    if ((Decl.StorageClass & SC_EXTERN) == 0) {
+    if ((Decl.StorageClass & SC_ALIAS) == 0 &&
+        (Decl.StorageClass & SC_EXTERN) == 0) {
         Decl.StorageClass |= SC_DEF;
     }
 
     /* Handle anything that needs storage (no functions, no typdefs) */
-    if ((Decl.StorageClass & SC_FUNC) != SC_FUNC &&
-         (Decl.StorageClass & SC_TYPEMASK) != SC_TYPEDEF) {
+    if ((Decl.StorageClass & SC_DEF) == SC_DEF &&
+        (Decl.StorageClass & SC_TYPEMASK) != SC_TYPEDEF) {
 
         /* If we have a register variable, try to allocate a register and
         ** convert the declaration to "auto" if this is not possible.
@@ -468,13 +469,6 @@ static void ParseOneDecl (const DeclSpec* Spec)
         } else if ((Decl.StorageClass & SC_AUTO) == SC_AUTO) {
             /* Auto variable */
             ParseAutoDecl (&Decl);
-        } else if ((Decl.StorageClass & SC_EXTERN) == SC_EXTERN) {
-            /* External identifier - may not get initialized */
-            if (CurTok.Tok == TOK_ASSIGN) {
-                Error ("Cannot initialize externals");
-            }
-            /* Add the external symbol to the symbol table */
-            AddLocalSym (Decl.Ident, Decl.Type, Decl.StorageClass, 0);
         } else if ((Decl.StorageClass & SC_STATIC) == SC_STATIC) {
             /* Static variable */
             ParseStaticDecl (&Decl);
@@ -483,6 +477,13 @@ static void ParseOneDecl (const DeclSpec* Spec)
         }
 
     } else {
+
+        if ((Decl.StorageClass & SC_EXTERN) == SC_EXTERN) {
+            /* External identifier - may not get initialized */
+            if (CurTok.Tok == TOK_ASSIGN) {
+                Error ("Cannot initialize externals");
+            }
+        }
 
         /* Add the symbol to the symbol table */
         AddLocalSym (Decl.Ident, Decl.Type, Decl.StorageClass, 0);

--- a/src/cc65/locals.c
+++ b/src/cc65/locals.c
@@ -426,13 +426,25 @@ static void ParseOneDecl (const DeclSpec* Spec)
     /* Read the declaration */
     ParseDecl (Spec, &Decl, DM_NEED_IDENT);
 
-    /* Set the correct storage class for functions */
+    /* Check if there are any non-extern storage classes set for function
+    ** declarations. The only valid storage class for function declarations
+    ** inside functions is 'extern'.
+    */
     if ((Decl.StorageClass & SC_FUNC) == SC_FUNC) {
-        /* Function prototypes are always external */
-        if ((Decl.StorageClass & SC_EXTERN) == 0) {
-            Warning ("Function must be extern");
+
+        /* Check if there are explicitly specified non-external storage classes */
+        if ((Spec->Flags & DS_DEF_STORAGE) != DS_DEF_STORAGE    &&
+            (Decl.StorageClass & SC_EXTERN) == 0                &&
+            (Decl.StorageClass & SC_STORAGEMASK) != 0) {
+            Error ("Illegal storage class on function");
         }
+
+        /* The default storage class could be wrong. Just use 'extern' in all
+        ** cases.
+        */
+        Decl.StorageClass &= ~SC_STORAGEMASK;
         Decl.StorageClass |= SC_EXTERN;
+
     }
 
     /* If we don't have a name, this was flagged as an error earlier.

--- a/src/cc65/main.c
+++ b/src/cc65/main.c
@@ -1069,7 +1069,7 @@ int main (int argc, char* argv[])
     /* Create the output file if we didn't had any errors */
     if (PreprocessOnly == 0 && (ErrorCount == 0 || Debug)) {
 
-        /* Emit literals, externals, do cleanup and optimizations */
+        /* Emit literals, do cleanup and optimizations */
         FinishCompile ();
 
         /* Open the file */

--- a/src/cc65/symtab.c
+++ b/src/cc65/symtab.c
@@ -52,13 +52,13 @@
 #include "declare.h"
 #include "error.h"
 #include "funcdesc.h"
+#include "function.h"
 #include "global.h"
+#include "input.h"
 #include "stackptr.h"
 #include "symentry.h"
 #include "typecmp.h"
 #include "symtab.h"
-#include "function.h"
-#include "input.h"
 
 
 
@@ -728,6 +728,10 @@ SymEntry* AddEnumSym (const char* Name, const Type* Type, SymTable* Tab)
         Entry->V.E.SymTab = Tab;
         Entry->V.E.Type   = Type;
 
+        if (Type != 0) {
+            Entry->Flags |= SC_DEF;
+        }
+
         /* Add it to the current table */
         AddSymEntry (CurTagTab, Entry);
     }
@@ -738,11 +742,12 @@ SymEntry* AddEnumSym (const char* Name, const Type* Type, SymTable* Tab)
 
 
 
-SymEntry* AddStructSym (const char* Name, unsigned Type, unsigned Size, SymTable* Tab)
+SymEntry* AddStructSym (const char* Name, unsigned Flags, unsigned Size, SymTable* Tab)
 /* Add a struct/union entry and return it */
 {
     SymTable* CurTagTab = TagTab;
     SymEntry* Entry;
+    unsigned Type = (Flags & SC_TYPEMASK);
 
     /* Type must be struct or union */
     PRECONDITION (Type == SC_STRUCT || Type == SC_UNION);
@@ -756,13 +761,14 @@ SymEntry* AddStructSym (const char* Name, unsigned Type, unsigned Size, SymTable
             /* Existing symbol is not a struct */
             Error ("Symbol '%s' is already different kind", Name);
             Entry = 0;
-        } else if (Size > 0 && Entry->V.S.Size > 0) {
+        } else if ((Entry->Flags & Flags & SC_DEF) == SC_DEF) {
             /* Both structs are definitions. */
             Error ("Multiple definition for '%s'", Name);
             Entry = 0;
         } else {
-            /* Define the struct size if it is given */
-            if (Size > 0) {
+            /* Define the struct size if it is a definition */
+            if ((Flags & SC_DEF) == SC_DEF) {
+                Entry->Flags      = Flags;
                 Entry->V.S.SymTab = Tab;
                 Entry->V.S.Size   = Size;
             }
@@ -777,7 +783,7 @@ SymEntry* AddStructSym (const char* Name, unsigned Type, unsigned Size, SymTable
     if (Entry == 0) {
 
         /* Create a new entry */
-        Entry = NewSymEntry (Name, Type);
+        Entry = NewSymEntry (Name, Flags);
 
         /* Set the struct data */
         Entry->V.S.SymTab = Tab;

